R/setkey.R

```diff
@@ -204,49 +204,41 @@ forderv = function(x, by=seq_along(x), retGrp=FALSE, sort=TRUE, order=1L, na.las
-  ans = x
-  order = rep(1L, length(cols))
-  if (length(cols)) {
-    ans = vector("list", length(cols))
-    cols = as.list(cols)
-    xcols = names(x)
-    for (i in seq_along(cols)) {
-      v=cols[[i]]
-      if (i == 1L && is.call(v) && length(v) == 2L && v[[1L]] == "list") return(1L) # to be consistent with base, see comment below under while loop
-      while (is.call(v) && length(v) == 2L && v[[1L]] != "list") {
-        # take care of "--x", "{-x}", "(---+x)" etc., cases and also "list(y)". 'list(y)' is ambiguous though. In base, with(DT, order(x, list(y))) will error
-        # that 'arguments are not of same lengths'. But with(DT, order(list(x), list(y))) will return 1L, which is very strange. On top of that, with(DT,
-        # order(x, as.list(10:1)) would return 'unimplemented type list'. It's all very inconsistent. But we HAVE to be consistent with base HERE.
-        if (!as.character(v[[1L]]) %chin% c("+", "-")) break   # FIX for bug #5583
-        if (v[[1L]] == "-") order[i] = -order[i]
-        v = v[[-1L]]
-      }
-      if (is.name(v)) {
-        ix = chmatch(as.character(v), xcols, nomatch=0L)
-        if (ix != 0L) ans = point(ans, i, x, ix) # see 'point' in data.table.R and C-version pointWrapper in assign.c - avoid copies
-        else {
-          v = as.call(list(as.name("list"), v))
-          ans = point(ans, i, eval(v, x, parent.frame()), 1L)
-        }
-      } else {
-        if (!is.object(eval(v, x, parent.frame()))) {
-          v   = as.call(list(as.name("list"), v))
-          ans = point(ans, i, eval(v, x, parent.frame()), 1L) # eval has to make a copy here (not due to list(.), but due to ex: "4-5*y"), unavoidable.
-        } else ans = point(ans, i, list(unlist(eval(v, x, parent.frame()))), 1L)
-      } # else stop("Column arguments to order by in 'forder' should be of type name/symbol (ex: quote(x)) or call (ex: quote(-x), quote(x+5*y))")
+  sub = substitute(list(...))
+  tt = sapply(sub, function(x) is.null(x) || (is.symbol(x) && !nzchar(x)))
+  if (any(tt)) sub[tt] = NULL  # remove any NULL or empty arguments; e.g. test 1962.052: forder(DT, NULL) and forder(DT, )
+  if (length(sub)<2L) return(NULL)  # forder() with no arguments returns NULL consistent with base::order
+  asc = rep.int(1L, length(sub)-1L)  # ascending (1) or descending (-1) per column
+  # the idea here is to intercept - before vectors so that :
+  # i) we can pass the decreasing argument for that column through to C to avoid what normally happens in R (allocate a new vector and apply - to every element first)
+  # ii) - on character vector works; ordinarily in R that fails with type error
+  # We intercept the unevaluated expressions and massage them before evaluating in with(DT) scope or not depending on the first item.
+  for (i in seq.int(2L, length(sub))) {
+    v = sub[[i]]
+    while (is.call(v) && length(v)==2L && ((s<-v[[1L]])=="-" || s=="+")) {
+      if (s=="-") asc[i-1L] = -asc[i-1L]
```

Comment:
why not just `-1L` on the RHS

Reply:
ah nvm. this handles `-+-+-+-` weird case

Reply:
yeah it seemed to be doing that already so I left it doing that